cmds/modprobe/modprobe.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// modprobe - Add and remove modules from the Linux Kernel
+//
+// Synopsis:
+//     modprobe modulename...
+//
+// Author:
+//     Roland Kammerer <dev.rck@gmail.com>
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/kmodule"
+	"golang.org/x/sys/unix"
+)
+
+const cmd = "modprobe modulename"
+
+type modState uint8
+
+const (
+	_ modState = iota
+	loading
+	loaded
+)
+
+type dependency struct {
+	state modState
+	deps  []string
+}
+
+type depMap map[string]*dependency
+
+var moduleDir = "/lib/modules/" // gets extended in main
+
+func genDeps(rd io.Reader) (depMap, error) {
+	deps := make(depMap)
+
+	scanner := bufio.NewScanner(rd)
+	for scanner.Scan() {
+		txt := scanner.Text()
+		nameDeps := strings.Split(txt, ":")
+		modPath, modDeps := nameDeps[0], nameDeps[1]
+		modPath = strings.TrimSpace(modPath)
+
+		var dependency dependency
+		if len(modDeps) > 0 {
+			for _, dep := range strings.Split(strings.TrimSpace(modDeps), " ") {
+				dependency.deps = append(dependency.deps, dep)
+			}
+		}
+		deps[modPath] = &dependency
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return deps, nil
+}
+
+func findModPath(modName string, m depMap) (string, error) {
+	for mp := range m {
+		if strings.HasSuffix(mp, modName+".ko") {
+			return mp, nil
+		}
+	}
+
+	return "", fmt.Errorf("modprobe: Could not find path for module %s", modName)
+}
+
+func loadModule(modPath, options string) error {
+	filename := filepath.Join(moduleDir, modPath)
+	f, err := os.Open(filename)
+	if err != nil {
+		return fmt.Errorf("modprobe: Could not open %q: %v", filename, err)
+	}
+	defer f.Close()
+
+	if err, e := kmodule.FileInit(f, options, 0); err != nil && e != unix.EEXIST {
+		return fmt.Errorf("modprobe: could not load %q: %v", filename, err)
+	}
+
+	return nil
+}
+
+func loadDeps(modPath string, m depMap) error {
+	dependency, ok := m[modPath]
+	if !ok {
+		return fmt.Errorf("modprobe: Could not find dependency %s", modPath)
+	}
+
+	if dependency.state == loading {
+		return fmt.Errorf("modprobe: Circular dependency, %s already in state LOADING", modPath)
+	} else if dependency.state == loaded {
+		return nil
+	}
+
+	m[modPath].state = loading
+
+	for _, dep := range dependency.deps {
+		if err := loadDeps(dep, m); err != nil {
+			return err
+		}
+	}
+
+	// done with dependencies, load module
+	if err := loadModule(modPath, ""); err != nil {
+		return err
+	}
+	m[modPath].state = loaded
+
+	return nil
+}
+
+func main() {
+	if len(os.Args) == 1 {
+		log.Fatalf("Usage: %s", cmd)
+	}
+
+	rel, err := exec.Command("uname", "-r").Output()
+	if err != nil {
+		log.Fatal(err)
+	}
+	moduleDir = filepath.Join(moduleDir, strings.TrimSpace(string(rel)))
```

Comment:
`/lib/modules/$(uname -r)/` is fairly standard. I've seen system's which skip the release number because they are guaranteed to only have one kernel. I haven't heard what the plan for u-root is yet.

Reply:
Then this needs feedback. I did what `modprobe(8)` told me, which is `$uname -r)`

Reply:
I'm not sure what the discussion here is -- this is using $(uname -r)? I'd do what modprobe(8) does.

Reply:
What you have now is good. We might decide to use a different path in the future.